lib/govuk_publishing_components/app_helpers/asset_helper.rb

```diff
@@ -0,0 +1,155 @@
+        hint
+        image-card
+        input
+        inset-text
+        intervention
+        inverse-header
+        label
+        layout-footer
+        layout-for-admin
+        layout-for-public
+        layout-header
+        layout-super-navigation-header
+        lead-paragraph
+        metadata
+        modal-dialogue
+        notice
+        organisation-logo
+        panel
+        phase-banner
+        previous-and-next-navigation
+        print-link
+        radio
+        related-navigation
+        reorderable-list
+        search
+        select
+        share-links
+        show-password
+        signup-link
+        single-page-notification-button
+        skip-link
+        step-by-step-nav
+        step-by-step-nav-header
+        step-by-step-nav-related
+        subscription-links
+        success-alert
+        summary-list
+        table
+        tabs
+        textarea
+        title
+        translation-nav
+        warning-text
+      ].map { |component|
+        "govuk_publishing_components/components/_#{component}.css"
+      }.freeze
+
+      def initialize
+        # Set up the lists of gem components, app components, and views.
+        @gem_component_stylesheets_being_used = []
+        @app_component_stylesheets_being_used = []
+        @view_stylesheets_being_used = []
+
+        # This list includes components already included in Static; taken from
+        # https://github.com/alphagov/static/blob/198a598682df40ce4a2c3c286c06244297c18cf0/app/assets/stylesheets/application.scss
+        #
+        # This is used to dedupe stylesheets without needing the full path.
+        @stylesheet_list = %w[
+          breadcrumbs
+          button
+          error-message
+          heading
+          hint
+          input
+          label
+          search
+          skip-link
+          textarea
+          title
+
+          cookie-banner
+          feedback
+          layout-footer
+          layout-for-public
+          layout-header
+          layout-super-navigation-header
+        ]
+      end
+
+      # Used to add a component that exists in the gem to the list
+      def add_gem_component(component)
+        unless is_already_used?(component)
+          @gem_component_stylesheets_being_used << "govuk_publishing_components/components/_#{component}.css"
+          @stylesheet_list << component
+        end
+      end
+
+      # Used to add a component that exists in the application to the list
+      def add_component(component)
+        unless is_already_used?(component)
+          @app_component_stylesheets_being_used << "components/_#{component}.css"
+          @stylesheet_list << component
+        end
+      end
+
+      # Some applications have view-specific stylesheets - use this method
+      # to add them to the list
+      def add_view(view)
+        unless is_already_used?(view)
+          @view_stylesheets_being_used << "views/_#{view}.css"
+          @stylesheet_list << view
+        end
+      end
```

Comment:
These 3 options seem a bit limiting, for example you couldn't have any stylesheets in a nested folder (e.g. view/dir-1/_partial.scss).

I'd imagine you'd do better with a generic method that takes a file path e.g: add_path("views/dir-1/_partial.css") which was called by the other convenience methods e.g.

```
def add_gem_component(component)
  add_path("govuk_publishing_components/components/_#{component}.css")
end
```

Reply:
Ah not quite, I meant introduce one method that accepts the relatively low level construct of the full path to an asset and that method has the business logic for dealing with a stylesheet being added.

Then any other methods are just wrappers around the low level method for convenience. 

So you'd end up with something like:

```
def add_gem_component(component)
  add_path("govuk_publishing_components/components/_#{component}.css")
end

def add_path(full_path)
   # all the business logic for adding a stylesheet
end
```

Reply:
Ah, and the `add_view_component` method can accept a whole path instead and pass this directly to `add_path`. Right. Thanks! Makes sense.

Reply:
Thanks, the methods have been updated so we can now support any file path, for example:

Generic method that takes any file path

```Ruby
      def add_stylesheet_path(component_path)
        unless is_already_used?(component_path)
          STYLESHEET_LIST << component_path
        end
      end
```

Convenience method that calls `add_stylesheet_path`

```Ruby
      def add_gem_component(gem_component)
        add_stylesheet_path("govuk_publishing_components/components/_#{gem_component}.css")
      end
```

Reply:
Now updated to make it clear that these are all stylesheets.